`pulumi convert` turns references to resources into `res.id` which is not valid in other languages

Creating the default `pulumi new serverless-aws-yaml` and then `pulumi convert --language typescript --out ./ts` succeeds - but then fails to typescheck in TS with:

```console
reviewing update (dev)

View Live: https://app.pulumi.com/lukehoban/demo2/dev/previews/158d94d9-3fcb-4d19-871c-0bf40984b85b

     Type                 Name       Plan     Info
     pulumi:pulumi:Stack  demo2-dev           1 error
 
Diagnostics:
  pulumi:pulumi:Stack (demo2-dev):
    error: Running program '/Users/lukehoban/dd/ced2022demos/demo2/demo2ts' failed with an unhandled exception:
    TSError: ⨯ Unable to compile TypeScript:
    index.ts(32,9): error TS2322: Type 'Output<string>' is not assignable to type 'Input<Function> | undefined'.
      Type 'Output<string>' is not assignable to type 'OutputInstance<Function>'.
        Types of property 'apply' are incompatible.
          Type '{ <U>(func: (t: string) => Promise<U>): Output<U>; <U>(func: (t: string) => OutputInstance<U>): Output<U>; <U>(func: (t: string) => U): Output<...>; }' is not assignable to type '{ <U>(func: (t: Function) => Promise<U>): Output<U>; <U>(func: (t: Function) => OutputInstance<U>): Output<U>; <U>(func: (t: Function) => U): Output<...>; }'.
            Types of parameters 'func' and 'func' are incompatible.
              Types of parameters 't' and 't' are incompatible.
                Type 'string' is not assignable to type 'Function'. 
```

Similar code is generated in other languages - which will similarly not typecheck correctly for them.

The issue is that this:

```yaml
eventHandler: ${fn}
```

Becomes this:

```ts
eventHandler: fn.id,
```

But should be this:

```ts
eventHandler: fn,
```

--- a/CHANGELOG_PENDING.md
+++ b/CHANGELOG_PENDING.md
@@ -29,3 +29,6 @@
 
 - Escaped interpolated strings now remove one extra dollar sign.
   [#382](https://github.com/pulumi/pulumi-yaml/pull/382)
+
+- Only insert "id" in ejected resource refs when the receiver type is a string.
+  [#389](https://github.com/pulumi/pulumi-yaml/pull/389)

--- a/pkg/pulumiyaml/codegen/load.go
+++ b/pkg/pulumiyaml/codegen/load.go
@@ -27,7 +27,6 @@ type importer struct {
 	referencedStacks []string
 
 	loader          pulumiyaml.PackageLoader
-	typing          pulumiyaml.Typing
 	configuration   map[string]*model.Variable
 	variables       map[string]*model.Variable
 	stackReferences map[string]*model.Variable
@@ -1074,16 +1073,9 @@ func (imp *importer) importTemplate(file *ast.TemplateDecl) (*model.Body, syntax
 
 // ImportTemplate converts a YAML template to a PCL definition.
 func ImportTemplate(file *ast.TemplateDecl, loader pulumiyaml.PackageLoader) (*model.Body, syntax.Diagnostics) {
-	// We are ignoring errors since runner will be treated as advisory
-	runner, _, _ := pulumiyaml.PrepareTemplate(file, nil, loader)
-	var typing pulumiyaml.Typing
-	if runner != nil {
-		// We ignore diags, since typing is advisory
-		typing, _ = pulumiyaml.TypeCheck(runner)
-	}
+
 	imp := importer{
 		loader:          loader,
-		typing:          typing,
 		configuration:   map[string]*model.Variable{},
 		variables:       map[string]*model.Variable{},
 		stackReferences: map[string]*model.Variable{},

--- a/pkg/pulumiyaml/codegen/load.go
+++ b/pkg/pulumiyaml/codegen/load.go
@@ -57,6 +57,12 @@ func (imp *importer) importRef(node ast.Expr, name string, environment map[strin
 		return model.VariableReference(v), nil
 	}
 	if v, ok := imp.resources[name]; ok {
+		// TODO: This comparison is not 100% sound, since hint may be a union that
+		// includes a `string` type but does not include the resource type itself.
+		//
+		// The solution is to expose IsAssignable from pulumiyaml.Typing, but for that to
+		// be reliable we would need to harden a lot of our top-sort algorithm against
+		// missing nodes.
 		exportID := codegen.UnwrapType(hint) == schema.StringType
 		if isAccess || !exportID {
 			return model.VariableReference(v), nil

--- a/pkg/pulumiyaml/codegen/load.go
+++ b/pkg/pulumiyaml/codegen/load.go
@@ -27,6 +27,7 @@ type importer struct {
 	referencedStacks []string
 
 	loader          pulumiyaml.PackageLoader
+	typing          pulumiyaml.Typing
 	configuration   map[string]*model.Variable
 	variables       map[string]*model.Variable
 	stackReferences map[string]*model.Variable
@@ -47,7 +48,7 @@ var null = &model.Variable{
 
 // importRef imports a reference to a config variable or resource. These entities all correspond to top-level variables
 // in the Pulumi program, so each reference is imported as a scope traversal expression.
-func (imp *importer) importRef(node ast.Expr, name string, environment map[string]model.Expression, isAccess bool) (model.Expression, syntax.Diagnostics) {
+func (imp *importer) importRef(node ast.Expr, name string, environment map[string]model.Expression, isAccess bool, hint schema.Type) (model.Expression, syntax.Diagnostics) {
 	// `pulumi` is not a real variable, so it doesn't make sense to look it up.
 	contract.Assertf(name != pulumiyaml.PulumiVarName, "%[1]T: %[1]v", node)
 	if v, ok := imp.configuration[name]; ok {
@@ -57,9 +58,11 @@ func (imp *importer) importRef(node ast.Expr, name string, environment map[strin
 		return model.VariableReference(v), nil
 	}
 	if v, ok := imp.resources[name]; ok {
-		if isAccess {
+		exportID := codegen.UnwrapType(hint) == schema.StringType
+		if isAccess || !exportID {
 			return model.VariableReference(v), nil
 		}
+
 		return &model.ScopeTraversalExpression{
 			Traversal: hcl.Traversal{hcl.TraverseRoot{Name: v.Name}, hcl.TraverseAttr{Name: "id"}},
 			Parts:     []model.Traversable{v, model.DynamicType},
@@ -117,7 +120,7 @@ func (imp *importer) pulumiPropertyAccess(node ast.Expr, accessors []ast.Propert
 	}
 }
 
-func (imp *importer) importPropertyAccess(node ast.Expr, access *ast.PropertyAccess, environment map[string]model.Expression) (model.Expression, syntax.Diagnostics) {
+func (imp *importer) importPropertyAccess(node ast.Expr, access *ast.PropertyAccess, environment map[string]model.Expression, hint schema.Type) (model.Expression, syntax.Diagnostics) {
 	var diags syntax.Diagnostics
 
 	accessors := access.Accessors
@@ -126,7 +129,7 @@ func (imp *importer) importPropertyAccess(node ast.Expr, access *ast.PropertyAcc
 		return f, diags
 	}
 
-	receiver, rdiags := imp.importRef(node, accessors[0].(*ast.PropertyName).Name, environment, len(accessors) > 1)
+	receiver, rdiags := imp.importRef(node, accessors[0].(*ast.PropertyName).Name, environment, len(accessors) > 1, hint)
 	diags.Extend(rdiags...)
 
 	traversal, parts := hcl.Traversal{}, []model.Traversable{model.DynamicType}
@@ -172,7 +175,7 @@ func (imp *importer) importInterpolate(node *ast.InterpolateExpr, substitutions 
 		parts = append(parts, plainLit(part.Text))
 
 		if part.Value != nil {
-			ref, rdiags := imp.importPropertyAccess(node, part.Value, environment)
+			ref, rdiags := imp.importPropertyAccess(node, part.Value, environment, nil)
 			diags.Extend(rdiags...)
 
 			parts = append(parts, ref)
@@ -470,7 +473,7 @@ func (imp *importer) importExpr(node ast.Expr, hint schema.Type) (model.Expressi
 		if f, ok, diags := imp.pulumiPropertyAccess(node, node.Property.Accessors); ok {
 			return f, diags
 		}
-		return imp.importPropertyAccess(node, node.Property, nil)
+		return imp.importPropertyAccess(node, node.Property, nil, hint)
 	case ast.BuiltinExpr:
 		return imp.importBuiltin(node)
 	default:
@@ -1071,8 +1074,16 @@ func (imp *importer) importTemplate(file *ast.TemplateDecl) (*model.Body, syntax
 
 // ImportTemplate converts a YAML template to a PCL definition.
 func ImportTemplate(file *ast.TemplateDecl, loader pulumiyaml.PackageLoader) (*model.Body, syntax.Diagnostics) {
+	// We are ignoring errors since runner will be treated as advisory
+	runner, _, _ := pulumiyaml.PrepareTemplate(file, nil, loader)
+	var typing pulumiyaml.Typing
+	if runner != nil {
+		// We ignore diags, since typing is advisory
+		typing, _ = pulumiyaml.TypeCheck(runner)
+	}
 	imp := importer{
 		loader:          loader,
+		typing:          typing,
 		configuration:   map[string]*model.Variable{},
 		variables:       map[string]*model.Variable{},
 		stackReferences: map[string]*model.Variable{},

--- a/pkg/tests/transpiled_examples/aws-static-website-pp/aws-static-website.pp
+++ b/pkg/tests/transpiled_examples/aws-static-website-pp/aws-static-website.pp
@@ -7,7 +7,7 @@ resource siteBucket "aws-native:s3:Bucket" {
 
 resource indexHtml "aws:s3/bucketObject:BucketObject" {
 	__logicalName = "index.html"
-	bucket = siteBucket.id
+	bucket = siteBucket
 	source = fileAsset("./www/index.html")
 	acl = "public-read"
 	contentType = "text/html"
@@ -15,7 +15,7 @@ resource indexHtml "aws:s3/bucketObject:BucketObject" {
 
 resource faviconPng "aws:s3/bucketObject:BucketObject" {
 	__logicalName = "favicon.png"
-	bucket = siteBucket.id
+	bucket = siteBucket
 	source = fileAsset("./www/favicon.png")
 	acl = "public-read"
 	contentType = "image/png"

